### PR TITLE
Ticket/18911

### PR DESCRIPTION
**Pull request by melinath**
This is targeted at [18911 - Use VideoFeed subclasses](http://bugzilla.pculture.org/show_bug.cgi?id=18911), but it also resolves the following tickets:
- [17246 - Importing Vimeo feed brings in fewer videos than the estimated video count](http://bugzilla.pculture.org/show_bug.cgi?id=17246)
- [18852 - generic feeds with links to YouTube don't pull in metadata](http://bugzilla.pculture.org/show_bug.cgi?id=18852)
- [18722 - Clean up suite API for fetching pages of a feed or search.](http://bugzilla.pculture.org/show_bug.cgi?id=18722)
- [18909 - Add BaseSuite.normalize_url method.](http://bugzilla.pculture.org/show_bug.cgi?id=18909)

### CODE REVIEW

File: vidscraper/__init__.py [hunk 1]
```diff
@@ -87,32 +81,43 @@ def auto_feed(url, fields=None, crawl=False, max_results=None, api_keys=None,
     :raises UnhandledURL: if no registered suites know how to handle this url.
 
     """
-    return VideoFeed(url, fields=fields, crawl=crawl, max_results=max_results,
-                       api_keys=api_keys, last_modified=last_modified,
-                       etag=etag)
+    for suite in registry.suites:
+        try:
+            return suite.get_feed(url,
```
Comment (paulswartz) [inline vidscraper/__init__.py]: Should we have a `registry.get_suite_for_feed_url` method, instead of iterating through all the suites?

File: vidscraper/suites/base.py [hunk 2]
```diff
@@ -241,42 +249,40 @@ def handles_video_url(self, url):
...
-        to do that conversion.  By default, this method just returns the
-        original URL.
-
-        """
-        return url
-
-    def get_feed(self, url, **kwargs):
-        """Returns a feed using this suite."""
-        return VideoFeed(url, self, **kwargs)
+            self.feed_class(url)
```
Comment (paulswartz) [inline vidscraper/suites/base.py]: This could probably be implemented as: 

``` python
try:
    self.get_feed(url)
    return True
except UnhandledURL:
    return False
```

for de-duplication's sake, but I don't think that's a dealbreaker.

File: vidscraper/suites/blip.py [hunk 3]
```diff
@@ -83,13 +54,71 @@ def get_url(self, video):
...
+
+        raise UnhandledURL(url)
+
+    def get_page_url_data(self, *args, **kwargs):
+        data = super(BlipFeed, self).get_page_url_data(*args, **kwargs)
+        show = self.url_data['show']
+        data['show_path'] = '{0}/'.format(show) if show is not None else ''
+        return data
+
+    def get_item_data(self, item):
```
Comment (paulswartz) [inline vidscraper/suites/blip.py]: `item` is pretty generic.  If we're renaming things anyways, can we rename them for clarity and make this `get_video_data`?

File: vidscraper/suites/vimeo.py [hunk 4]
```diff
@@ -113,17 +116,392 @@ def process(self, response):
...
+        if not self.has_api_keys():
+            raise UnhandledSearch
+        # Vimeo's search api supports relevant, newest, oldest, most_played,
+        # most_commented, and most_liked. We only support relevant and newest,
+        # for now.
+        if self.order_by == 'relevant':
+            pass
+        elif self.order_by == 'latest':
+            self.order_by = 'newest'
+        else:
```
Comment (paulswartz) [inline vidscraper/suites/vimeo.py]: Maybe pass the other types along to Vimeo, if we know they're supported? We say we support 'popular' as a type, so that would seem to map to 'most_played'.

File: vidscraper/suites/youtube.py [hunk 5]
```diff
@@ -185,120 +148,169 @@ def process(self, response):
...
+class YouTubeSearch(VideoSearch):
+    per_page = 50
+    page_url_format = ('http://gdata.youtube.com/feeds/api/videos?v=2&alt=json&'
+                       'q={query}&orderby={order_by}&start-index={page_start}&'
+                       'max-results={page_max}')
+
+    def __init__(self, query, order_by='relevant', **kwargs):
+        super(YouTubeSearch, self).__init__(query, order_by, **kwargs)
+        # YouTube's search supports relevance, published, viewCount, and
+        # rating. We currently only support relevance and published.
```
Comment (paulswartz) [inline vidscraper/suites/youtube.py]: We say we support 'popular', which would map to 'viewCount'.

File: vidscraper/__init__.py (hunk 1 above)
Comment (melinath) [inline vidscraper/__init__.py]: Since that method would also need to iterate through the suites, it doesn't seem like there would be a clear benefit to that.

Comment (melinath) [inline vidscraper/__init__.py]: Actually, there is already a method for that; in one sense, this is duplication, since we could use `suite_for_feed_url` and then `get_feed` from that - but since the `suite_for_feed_url` method calls `suite.handles_feed_url`, which calls `suite.get_feed`, implementing it this way is more direct... still, this could be a good place to refactor some more.

File: vidscraper/suites/vimeo.py (hunk 4 above)
Comment (melinath) [inline vidscraper/suites/vimeo.py]: In general, I think that's a good idea, but it would also be easy to do as a separate ticket. Where do we say we support 'popular'?

File: vidscraper/suites/blip.py (hunk 3 above)
Comment (melinath) [inline vidscraper/suites/blip.py]: This method maps approximately to the old `parse_feed_entry` method. I chose `item` because I wanted to emphasize that the thing the data was coming from didn't have to be any particular form - could be a feedparser entry, json, html, w/e. But `get_video_data` might be better - emphasizes where the data is headed, instead of where it's coming from; really, the naming I was using should've been `get_data_from_item`, which would cumbersome.

tldr: good idea.

File: vidscraper/__init__.py (hunk 1 above)
Comment (melinath) [inline vidscraper/__init__.py]: The `suite_for_feed_url` method actually also has the fallback, so we should probably be using that.

File: vidscraper/suites/vimeo.py (hunk 4 above)
Comment (paulswartz) [inline vidscraper/suites/vimeo.py]: https://github.com/pculture/vidscraper/blob/develop/vidscraper/videos.py#L405

**Comment by melinath** [comment]
These commits should address your suggestions.

**Comment by paulswartz** [comment]
Looks good. I opened http://bugzilla.pculture.org/show_bug.cgi?id=19006 for the bare exception raises because I don't think it's necessary to fix as part of this ticket.
